### ISSUE TITLE
How do I do string comparison in BPF programs ?

### ISSUE THREAD

**Issue by chenhengqi** [iovisor/bcc#3559]
I tried `__builtin_memcmp`, it compiles, but libbpf complains:

```
libbpf: failed to find BTF for extern 'memcmp': -2
Error: failed to open BPF object file: No such file or directory
```

Though I can use a bounded loop to bypass this, I want to know if there is a graceful solution. Thanks.

**Comment by yonghong-song** [comment]
Yes, for non-constant memcmp length, llvm won't convert __builtin_memcmp to a loop. So your bounded loop approach is a common practice for now. In the past, there are some discussions to introduce some bpf helpers for string/memory operations (e.g, strcmp or memcmp), since you have such a use case, maybe you can propose?

**Comment by chenhengqi** [comment]
Thanks. May have a try later.

**Comment by Wqrld** [comment]
> Thanks. May have a try later.

Were you able to get a functioning bounded loop together with the same function as memcmp? I have tried but keep rejected by the verifier for out of bounds accesses while having bound checks in place.

**Comment by chenhengqi** [comment]
> > Thanks. May have a try later.
> 
> Were you able to get a functioning bounded loop together with the same function as memcmp? I have tried but keep rejected by the verifier for out of bounds accesses while having bound checks in place.

Yes. It works for me.

**Comment by ghostpepper108** [comment]
For folks who may need it, here is a code snipper that I found in one of the examples

static int local_strcmp(const char *cs, const char *ct)
{
    int len = 0;
    unsigned char c1, c2;

    while (len++ < sizeof(ct))
    {
        c1 = *cs++;
        c2 = *ct++;

        if (c1 != c2)
            return c1 < c2 ? -1 : 1;

        if (!c1)
            break;
    }
    return 0;
}

**Comment by chenhengqi** [comment]
> For folks who may need it, here is a code snipper that I found in one of the examples
> 
> static int local_strcmp(const char *cs, const char *ct)
> {
> int len = 0;
> unsigned char c1, c2;
> 
> ```
> while (len++ < sizeof(ct))
> {
>     c1 = *cs++;
>     c2 = *ct++;
> 
>     if (c1 != c2)
>         return c1 < c2 ? -1 : 1;
> 
>     if (!c1)
>         break;
> }
> return 0;
> ```
> 
> }

It's not right to use `sizeof` here.

**Comment by chenhengqi** [comment]
> Were you able to get a functioning bounded loop together with the same function as memcmp? I have tried but keep rejected by the verifier for out of bounds accesses while having bound checks in place.

I use this to compare task comm.
```c
static bool strequal(const char *a, const char *b)
{
	int i;

	for (int i = 0; i < 16; i++) {
		if (a[i] != b[i])
			return false;
	}
	return true;
}
```

**Comment by cdjianghan** [comment]
Is there any way to do string compare without loop？My kernel version is 4.14 which can‘t support loop。

**Comment by cdjianghan** [comment]
thx ,I have no choice, the kernel version in our cluster is not my call。

**Comment by chenhengqi** [comment]
Have you tried `#pragma unroll` ?

**Comment by cdjianghan** [comment]
anyway thx ,I will try #pragma  unroll,although I have failed before.
